src/ui/public/vis_maps/visualizations/marker_types/base_marker.js

```diff
@@ -229,8 +230,51 @@ export default function MarkerFactory() {
       this._createTooltip(content, latLng);
     }
 
+    _getOffset(content, latLng) {
+      // Default maxWidth leaflet applies to popup is 300
+      const popupWidth = 300;
+
+      // We need to create the popup first to determine how tall it will be.  Give
+      // it an out of map bounds offset as we don't want to see this one.  It will
+      // get replaced on the next popup creation where we apply the correct offset
+      const popup = L.popup({ autoPan: false, offset: new L.Point(this.map.getSize().x * 2, this.map.getSize().y * 2) })
+      .setLatLng(latLng)
+      .setContent(content)
+      .openOn(this.map);
+      const popupHeight = popup._contentNode.clientHeight;
+
+      const east = this.map.getBounds().getEast();
+      const west = this.map.getBounds().getWest();
+      const north = this.map.getBounds().getNorth();
+      const south = this.map.getBounds().getSouth();
+      const width = Math.abs(east - west);
+      const height = Math.abs(north - south);
+      const xscale = this.map.getSize().x / width;
+      const yscale = this.map.getSize().y / height;
+
+      const popupDistanceToLeftEdge = Math.abs(latLng.lng - west) * xscale;
+      const popupDistanceToRightEdge = Math.abs(latLng.lng - east) * xscale;
+      let widthOffset = 0;
+      if (popupDistanceToLeftEdge < popupWidth / 2) {
+        widthOffset = popupWidth / 2;
```

Comment:
for the pop-up to attach to the point, we'd probably want:
`widthOffset = popupWidth /  4 + 16;`

![image](https://cloud.githubusercontent.com/assets/1833023/22716185/addfd0fa-ed62-11e6-998d-9622b0922a18.png)

Similar for right side:
`widthOffset = -popupWidth / 4 - 16`;